### PR TITLE
`TokenStream`-based attributes, paths in attribute and derive macro invocations

### PR DESCRIPTION
**Pull request by jseyfried**
This PR
 - refactors `Attribute` to use  `Path` and `TokenStream` instead of `MetaItem`.
 - supports macro invocation paths for attribute procedural macros.
   - e.g. `#[::foo::attr_macro] struct S;`, `#[cfg_attr(all(), foo::attr_macro)] struct S;`
 - supports macro invocation paths for derive procedural macros.
   - e.g. `#[derive(foo::Bar, super::Baz)] struct S;`
 - supports arbitrary tokens as arguments to attribute procedural macros.
   - e.g. `#[foo::attr_macro arbitrary + tokens] struct S;`
 - supports using arbitrary tokens in "inert attributes" with derive procedural macros.
   - e.g. `#[derive(Foo)] struct S(#[inert arbitrary + tokens] i32);`
where `#[proc_macro_derive(Foo, attributes(inert))]`

r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @keeperofdakeys @abonander @SimonSapin @cgswords @dtolnay @Sergio

**Comment by dtolnay** [comment]
What is the use case that motivates arbitrary tokens in attributes?

The test case that has the following is not a real thing that we expect people to write, correct?

```rust
#[::attr_args::identity
  fn main() { assert_eq!(foo(), "Hello, world!"); }]
struct Dummy;
```

**Comment by jseyfried** [comment]
@dtolnay
> What is the use case that motivates arbitrary tokens in attributes?

We want to give attribute procedural macro authors the power/flexibility to parse the attributes however they want. Also, we pass the attribute's arguments to the procedural macro as a `TokenStream`, so it makes sense to support any `TokenStream`.

> The test case that has the following is not a real thing that we expect people to write, correct?

Well, it's a not a useful procedural macro, but people can put whatever they want in macro attributes. Items in macro attributes is technically fine but looks ugly so I wouldn't expect it in the wild. However, it seems reasonable to for people to use other AST nodes (e.g. expressions, patterns, paths, etc.) in attribute macros.

**Comment by dtolnay** [comment]
I guess I would strongly prefer to have a concrete use case in mind to motivate this change. Are there any existing procedural macro crates whose API you think would be improved by using "unconventional" attributes? Are there any procedural macro crates that you wish could exist but require this feature?

I agree that we can come up with hypothetical ways to use this but so far I disagree that this functionality is the best way to address any of those use cases, and consequently my instinct would be to discourage libraries from using this.

I don't find "we pass the attribute's arguments to the procedural macro as a TokenStream" to be a compelling justification. The only thing that means is that it is technically possible to implement this change, not that it is a good idea.

**Comment by nrc** [comment]
@dtolnay an example might be an attribute on an enum that wants to do something specific with one field:

```
#[foo::macro(Var2::field2)]
enum Bar {
    Var1,
    Var2 { field1: i32, field2: i32 }
}
```

or maybe specifying a precondition:

```
#[precondition x < 100]
fn foo(x: i32) { ... }
```

**Comment by nrc** [comment]
Concretely, my libhoare crate does the precondition thing, but hacks it with a string atm.

**Comment by dtolnay** [comment]
Neat, the libhoare use case is compelling. Thanks!

**Comment by jseyfried** [comment]
@dtolnay Also, the "unconventional" attributes this PR allows are gated behind `feature(proc_macro)`, so we can always decide to change course and add some restrictions later.

File: src/librustc/middle/stability.rs [hunk 1]
```diff
@@ -402,7 +402,8 @@ impl<'a, 'tcx> Index<'tcx> {
 
         let mut is_staged_api = false;
         for attr in &krate.attrs {
-            if attr.name() == "stable" || attr.name() == "unstable" {
+            let name = match attr.name() { Some(name) => name, _ => continue };
```
Comment (nrc) [inline src/librustc/middle/stability.rs]: worth factoring this idiom out into a macro? Maybe doesn't get used enough, but I'm not really a fan of match one-liners

File: src/librustc_resolve/macros.rs [hunk 2]
```diff
@@ -194,9 +197,11 @@ impl<'a> base::Resolver for Resolver<'a> {
 
         // Check for legacy derives
         for i in 0..attrs.len() {
-            if attrs[i].name() == "derive" {
+            let name = match attrs[i].name() { Some(name) => name, None => continue };
```
Comment (nrc) [inline src/librustc_resolve/macros.rs]: Ha, 4 uses, definitely needs a macro :-)

File: src/libsyntax/attr.rs [hunk 3]
```diff
@@ -185,26 +189,38 @@ impl NestedMetaItem {
...
+    pub fn meta_item_list(&self) -> Option<Vec<NestedMetaItem>> {
+        match self.meta() {
+            Some(MetaItem { node: MetaItemKind::List(list), .. }) => Some(list),
+            _ => None
+        }
     }
 
-    pub fn is_word(&self) -> bool { self.meta().is_word() }
+    pub fn is_word(&self) -> bool {
+        self.tokens.is_empty()
```
Comment (nrc) [inline src/libsyntax/attr.rs]: This seems a bit odd to me - this basically means that the attribute is just a path with no other tokens? But that means `is_word` is a bad name - either there should be a check that the path has length 1 or the name should change.

**Review by nrc — APPROVED** [review]
r=me with the naming and refactoring addressed

File: src/libsyntax/attr.rs (hunk 3 above)
Comment (jseyfried) [inline src/libsyntax/attr.rs]: Good point -- I'll add a check that the path has length 1.

File: src/librustc_resolve/macros.rs (hunk 2 above)
Comment (jseyfried) [inline src/librustc_resolve/macros.rs]: Will do.

**Comment by alexcrichton** [comment]
@bors: retry

actual error:

```
curl: (7) Failed to connect to www.openssl.org port 443: Connection refused


command did not execute successfully: "curl" "-o" "/checkout/obj/build/mipsel-unknown-linux-gnu/openssl/openssl-1.0.2k.tar.tmp" "https://www.openssl.org/source/openssl-1.0.2k.tar.gz"
expected success, got: exit code: 7
```

* spurious network failure https://github.com/rust-lang/rust/issues/40474

**Comment by matklad** [comment]
Hm, should arbitrary token trees in attributes require an RFC? At least, there was an RFC for the current `meta_item` stuff and arbitrary literals.

I would feel slightly more confident if attributes were restricted to, say expressions. The problem that I see is that currently, `#[foo(bar, baz quuz spam)]` is a syntax error, which is easily detected during parsing. With token trees, this is a totally valid syntax though, and you'd need *semantic* knowledge just to parse this thing, which hurts IDEs.

I don't think I necessary disagree with allowing arbitrary token trees, but it does make me feel uneasy :)

**Comment by matklad** [comment]
Found one more interesting problem with arbitrary token trees in attributes: they make good error recovery very challenging. 

Consider, for example, this code:

```Rust
#[cfg(foo,
fn foo() {
}

mod m {
}
```

With the previous restricted grammar, the parser, after consuming `foo,` and waiting for the next meta item, would see a `fn` token, which can not start a meta item, but belongs to the `FIRST(item)`. So the natural and easy thing to do would be to report an error, stop parsing the attributes, and then proceed with parsing the following function and module. 


Now, when arbitrary token trees are allowed in attributes, the parser has to parse all of the following code as a part of an attribute, and so it would not recognize than a function and a module are, in fact, a function and a module.

Granted, the similar effect is also present with Rust macros, but it is expected that macros are annoying to work with :-)

**Pull request merged**